File: ios/Konnecta/Data/Remote/LeaderboardService.swift
import Foundation
import Supabase

struct LeaderboardEntry: Codable, Identifiable {
    var id: String { email }
    val full_name: String?
    val avatar_url: String?
    val email: String
    val visit_count: Int
}

private struct PlanWithProfileOnly: Codable {
    val user_id: String
    val profiles: Profile
}

class LeaderboardService {
    private let client = SupabaseManager.shared.client
    
    func getLeaderboard(groupId: String) async -> [LeaderboardEntry] {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let today = dateFormatter.string(from: Date())
        
        do {
            let plans: [PlanWithProfileOnly] = try await client.database
                .from("weekend_plans")
                .select(columns: "user_id, profiles(id, full_name, avatar_url, email, updated_at)")
                .eq(column: "status", value: "going")
                .eq(column: "group_id", value: groupId)
                .lt(column: "weekend_date", value: today)
                .execute()
                .value
            
            let grouped = Dictionary(grouping: plans, by: { $0.user_id })
            let entries = grouped.map { (userId, userPlans) -> LeaderboardEntry in
                let profile = userPlans.first!.profiles
                return LeaderboardEntry(
                    full_name = profile.full_name,
                    avatar_url = profile.avatar_url,
                    email = profile.email,
                    visit_count = userPlans.count
                )
            }
            
            return entries.sorted(by: { $0.visit_count > $1.visit_count }).prefix(5).map { $0 }
        } catch {
            print("Leaderboard error: \(error)")
            return []
        }
    }
}

File: android/app/src/main/kotlin/com/konnecta/app/data/remote/LeaderboardService.kt
package com.konnecta.app.data.remote

import com.konnecta.app.data.model.*
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LeaderboardService @Inject constructor() {
    private val client = SupabaseClient.client

    suspend fun getLeaderboard(groupId: String): List<LeaderboardEntry> {
        val today = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())

        return try {
            val plans = client.postgrest["weekend_plans"]
                .select(columns = Columns.raw("user_id, profiles(id, full_name, avatar_url, email, updated_at)")) {
                    filter {
                        eq("status", "going")
                        eq("group_id", groupId)
                        lt("weekend_date", today)
                    }
                    limit(500)
                }
                .decodeList<PlanWithProfileOnly>()

            plans.groupBy { it.user_id }
                .map { (_, userPlans) ->
                    val profile = userPlans.first().profiles
                    LeaderboardEntry(
                        full_name = profile.full_name,
                        avatar_url = profile.avatar_url,
                        email = profile.email,
                        visit_count = userPlans.size
                    )
                }
                .sortedByDescending { it.visit_count }
        } catch (e: Exception) {
            Timber.e(e, "LeaderboardService: Error fetching leaderboard")
            emptyList()
        }
    }
}
